File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left, right = 0, len(height) - 1
        max_left, max_right = 0, 0
        res = 0
        while(left <= right):
            max_left = max(max_left, height[left])
            max_right = max(max_right, height[right])
            min_h = min(max_left, max_right)
            if max_left == min_h:
                res += min_h - height[left]
                left = left + 1
            else:
                res += min_h - height[right]
                right = right - 1
        return res
        

File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int N = height.length;
        int left = 0, right = N - 1;
        int leftMax = 0, rightMax = 0;
        int ans = 0;
        while(left < right){
            if(height[left] < height[right]){
                //说明left一定能储水
                if(height[left] >= leftMax){
                    leftMax = height[left];
                }else{
                    ans += (leftMax - height[left]);
                }
                left++;
            }else{
                //说明right一定能储水
                if(height[right] >= rightMax){
                    rightMax = height[right];
                }else{
                    ans += (rightMax - height[right]);
                }
                right--;
            }
        }
        return ans;

    }

}